ynr/apps/candidates/tests/test_models.py

```diff
@@ -27,15 +28,22 @@ def test_get_sopn_view(self):
 
 class TestBallotMethods(TestCase):
     def test_uncontested(self):
-        ballot = self.Ballot(
-            ballot_paper_id="local.sheffield.ecclesall.2021-05-06"
+        contested_ballot = Ballot(
+            ballot_paper_id="local.sheffield.ecclesall.2021-05-06",
+            winner_count=3,
+            candidates_locked=False,
+        )
+        uncontested_ballot = Ballot(
+            ballot_paper_id="local.sheffield.ecclesall.2021-05-06",
+            # winner_count=3,
```

Comment:
I'm not sure how to mock `membership_set_count` here

Reply:
Rather than mock it, it might be easier to actually create some `Membership` objects (the candidates standing). As from experience mocking a reverse FK relationship on a model gets a bit 🤯 

So what you will need is a saved Ballot object with x numbers of candidates. Then one test where the number of candidates > than winner count, and another where they are the same.

There are some helper mixins that you should be able to use to help with the creation of `Ballot`, `Person` and `Membership` objects you need. The one I have been using in some other tests is the [SingleBallotStatesMixin](https://github.com/DemocracyClub/yournextrepresentative/blob/cd4d78435221bcfa25be9f88e9874611eb7608f2/ynr/apps/elections/tests/test_ballot_view.py#L34) as it has some helper methods on it to create elections, ballots, memberships etc. So if you import that class and update your test class to inherit it, then you can use those methods to help set up objects for your tests. For example:
```
class TestBallotMethods(SingleBallotStatesMixin, TestCase):

    def setUp(self):
        # create the base objects we need - 
        election = self.create_election(name="Sheffield Local Election")
        post = self.create_post(post_label="Ecclesall")
        # set winner_count to 1 and make sure candidates locked
        self.ballot = self.create_ballot(election=election, post=post, winner_count=1, candidates_locked=True)


    def test_uncontested_when_winner_count_same_as_memberships(self):
        # create a single candidate for this case
        party = PersonFactory()
        person = PersonFactory()
        MembershipFactory(ballot=self.ballot, person=person, party=party)

        # so now we have 1 candidate so when we call ballot.uncontested,
        # the line in the method self.membership_set.count() will count
        # the numeber of membership objects and return 1 so...
        self.assertTrue(self.ballot.uncontested)

```

Then you could follow the above pattern but create multiple candidates - and then expect the result to be `False`. And possibly even another example where winner count is greater than, but candidates number still matches it (e.g. winner count 3, with 3 Memberships created) and still expect it to be uncontested

Reply:
Are you creating more than one Election in your test? As this may be due to how the ElectionFactory is setup. It inherits from [BaseElectionFactory](https://github.com/DemocracyClub/yournextrepresentative/blob/cd4d78435221bcfa25be9f88e9874611eb7608f2/ynr/apps/candidates/tests/factories.py#L31) and the slug hardcoded using the election date. So if you try to create a second election, unless a different `slug` or `election_date` is used, it will try to create another object with the same slug - which needs to be unique, so you get the error.

So if you are creating multiple `Election` objects in the same test, you will need to pass in a different slug for subsequent objects. Although saying that, I don't think you actually need more than one Election - just different `Ballot` objects assigned to a single `Election`